x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java

```diff
@@ -230,9 +272,86 @@ static void getMemoryAndCpu(
                 extraSingleNodeProcessors,
                 extraModelMemoryInBytes,
                 extraProcessors,
-                0,
+                removeNodeMemoryInBytes,
                 MachineLearning.NATIVE_EXECUTABLE_CODE_OVERHEAD.getBytes()
             )
         );
     }
+
+    /**
+     * Try to remove node memory
+     *
+     * @param perNodeUsedModelMemoryInBytes per Node lists of used model memory
+     * @param perNodeMemoryInBytes total model memory available on every node
+     * @return true if a node can be removed, false if not
+     */
+    static boolean tryRemoveOneNode(Map<String, List<Long>> perNodeUsedModelMemoryInBytes, long perNodeMemoryInBytes) {
+        if (perNodeUsedModelMemoryInBytes.size() <= 1) {
+            return false;
+        }
+
+        Map<String, Long> perNodeModelMemoryInBytesSum = perNodeUsedModelMemoryInBytes.entrySet()
+            .stream()
+            .map(entry -> tuple(entry.getKey(), entry.getValue().stream().mapToLong(Long::longValue).sum()))
+            .collect(Collectors.toMap(Tuple::v1, Tuple::v2));
+
+        Optional<Map.Entry<String, Long>> leastLoadedNodeAndMemoryUsage = perNodeModelMemoryInBytesSum.entrySet()
+            .stream()
+            .min(Comparator.comparingLong(Map.Entry::getValue));
+
+        if (leastLoadedNodeAndMemoryUsage.isPresent() == false) {
+            return false;
+        }
+
+        // this currently only works based on memory, not CPU
```

Comment:
In the case of trained models with high numbers of threads configured this will cause scaling loops, so needs to be fixed before merging.

Reply:
I thought so, too. But when I looked into the assignment code, I couldn't find logic that stops assignment if the number of CPU's doesn't fit with threads. But it is a complex code base, maybe I haven't found it.

There is one case for sure: The settings have a max number of jobs per node, that's not taken into account here as well.

Reply:
To be precise: The planer takes cores into account and tries to find a good fit, but:

- it doesn't stop assignment it seems
- it doesn't sum the number of taken threads/cores per node

Reply:
thinking about it a little more, it makes sense to handle this independently whether or not the planner takes CPU as a hard requirement or not. I might be added later and we don't want surprises due to assumptions. 

I will implement the same way as we handle upscale. I will also add a check for the max jobs per node limit.

Reply:
> it makes sense to handle this independently whether or not the planner takes CPU as a hard requirement or not

Yes, I agree.

You're right that the planner will do the best it can given insufficient resources. The planner was designed for on-prem, where this situation could arise.

The autoscaler can make things better by making sure there is enough space for all allocations. Then, when the planner runs after autoscaling completes it _should_ be able to allocate all requested allocations.

I think the comment "this currently only works based on memory, not CPU" can be removed now.